### ISSUE TITLE
[Security] Supply chain vulnerabilities in Meshery dependences: container images and libraries

### ISSUE THREAD

**Issue by leecalcote** [meshery/meshery#4642]
#### Current Behavior
See this [security score report](https://artifacthub.io/packages/helm/meshery/meshery?modal=security-report) for details of known vulnerabilities in Meshery's published packages.

#### Desired Behavior
Ideally, Meshery's security report is an A+.

#### Implementation
Volunteers needed to pick specific vulnerabilities and address them.

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and Resources
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

**Comment by alphaX86** [comment]
@leecalcote IG the security vulnerabilities arises from the OS image used, say like if I used Ubuntu as OS image by default it has a vulnerability from it's system package... 

But anyways, I'd like to volunteer on fixing this one

**Comment by alphaX86** [comment]
@leecalcote IG the security vulnerabilities arises from the OS image used, say like if I used Ubuntu as OS image by default it has a vulnerability from it's system package... 

But anyways, I'd like to volunteer on fixing this one

**Comment by Aisuko** [comment]
https://github.com/GoogleContainerTools/distroless Hi, guys. Shall we think about this project as the base image?

**Comment by alphaX86** [comment]
This looks neat 👍 🔥 so does that mean we're going to replace all Docker images with GCR Distroless images instead of Alpine image?

**Comment by Aisuko** [comment]
> This looks neat 👍 🔥 so does that mean we're going to replace all Docker images with GCR Distroless images instead of Alpine image?

Hi @alphaX86 , you are right here. We can replace the adapters base images due to adapters with the sample logic may not have extra tools to debug. But for Meshery main server, we need to consider.

**Comment by Aisuko** [comment]
I believe we should fix the secure issue of the base image. Remove the tool we don't need and make the image size as small as it could be(I believe this is our philosophy @leecalcote ). And please releate the PR to this issue and the list below we need to fix. I will try use the distroless container tool on the adapters.


- [ ] meshery/meshery:stable-latest
- [ ] meshery/meshery-cpx:stable-latest
- [ ] meshery/meshery-nsm:stable-latest
- [ ] meshery/meshery-osm:stable-latest
- [ ] meshery/meshery-kuma:stable-latest
- [ ] meshery/meshery-istio:stable-latest
- [ ] meshery/meshery-consul:stable-latest
- [ ] meshery/meshery-linkerd:stable-latest
- [ ] meshery/meshery-app-mesh:stable-latest
- [ ] meshery/meshery-nginx-sm:stable-latest
- [ ] meshery/meshery-operator:stable-latest
- [ ] gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0
- [ ] meshery/meshery-traefik-mesh:stable-latest

**Comment by Aisuko** [comment]
https://github.com/meshery/meshery-linkerd/pull/251

**Comment by alphaX86** [comment]
![image](https://user-images.githubusercontent.com/59508546/143047685-a22649fb-a754-465c-abab-7818f03cf440.png)

Seems the vulnerabilities arises from the system packages (like the above example), so does that mean updating the image to latest one will fix the above? Or would it be like updating the packages altogether? @Aisuko

**Comment by Aisuko** [comment]
Yes, you are right here. We may need to update the base image which we're using:

* Alpine
* Debian 10.9 
* Node.js

But If we can make a choice on all the adapter migrate to the seem base image e.g: like Debian 11. I believe we can fix the system-base image security issue easily. Like the operator 's base image.


<img width="651" alt="Screenshot 2021-11-24 at 7 17 00 AM" src="https://user-images.githubusercontent.com/8053949/143144195-d4b3c997-9f3f-437e-b2f3-c3d096c11461.png">

**Comment by alphaX86** [comment]
Is there any other way to get a security report for the container images? IG for Helm, it'd be done only when a new version is released in the Hub...

**Comment by Aisuko** [comment]
Maybe DockerHub? I'm not sure if we deploy a local Harbor(with the image scan features at local) can help us.

**Comment by alphaX86** [comment]
Dockerhub.... It's fine, it also has a scanner but a premium feature though :') I'm not sure if we have it.

Harbor... I'll give look on it

**Comment by alphaX86** [comment]
One common CVE is critical named... CVE-2021-33574 IDK if there's a way to solve it...

**Comment by alphaX86** [comment]
Bumping the tags solved 70% of CVEs (all high and critical range, some low kind still remain)

**Comment by leecalcote** [comment]
Wow!

**Comment by alphaX86** [comment]
I just had a doubt.... did we implement the Snyk vulnerabilty scanner in meshery org?

**Comment by leecalcote** [comment]
@alphaX86 let's get you access to have a look at the current set of scans more deeply.

**Comment by vishal267** [comment]
We can use trivy for scanning images, better to add in GitHub actions. Let me know if I can pick up this issue. Can aim to fix critical vulnerabilities. @leecalcote 

trivy image --severity CRITICAL  meshery/meshery-cpx:stable-latest

<img width="1021" alt="image" src="https://user-images.githubusercontent.com/19798194/188977842-355df8bc-2a0b-4754-9931-7d67c20ae4f6.png">

**Comment by leecalcote** [comment]
@vishal267 yes, let's give this a shot. Our challenge isn't so much knowing about the vulns, but in fixing them. We need someone to wrangle the updating of images, dependencies, and so on.

**Comment by leecalcote** [comment]
@vishal267 what do you propose for next steps?

**Comment by vishal267** [comment]
We can upgrade the base image in the docker file to "golang:1.19.1-alpine3.16". Checked with trivy scanner there is no critical vulnerability. FYI currently we are using  golang:1.17. Using an alpine image will help in reducing the image size. 
@leecalcote

**Comment by leecalcote** [comment]
Good   I suspect you’ll find compatibility issues with the Alpine base image and the embedded wrk2 and nighthawk binaries. If not, that would be great.

**Comment by acald-creator** [comment]
I checked the `Go` version and saw that `1.17` is EOL.

Here is the reference: https://endoflife.date/go

Perhaps we can start a slow process to upgrade the Go version to 1.18, and then we can another move to Go version 1.19

**Comment by leecalcote** [comment]
@acald-creator sounds great, yes, let's do it. In fact, we can start with 1.19, and if no issues arise, we might leapfrog 1.18. If issues do arise, we can bump down to 1.18 and assess whether the same concern(s) is present.

**Comment by acald-creator** [comment]
@leecalcote I have a separate branch for `meshery` repo on go 1.18 so if there were many differences in between 1.17 and 1.18. I'll create another branch to see if 1.19 is not too bad.

**Comment by leecalcote** [comment]
Oh, ok. Sounds great. Sounds like progress already. 🚂

**Comment by leecalcote** [comment]
I should mention just to be sure that it’s part of the plan: all of the components should be upgraded concurrently.

**Comment by acald-creator** [comment]
Agreed, I'll raise the issue but this issue can be a tracking for all components, and we can make references to the other repos as we goes through the changes.

**Comment by leecalcote** [comment]
Oh, no. Did the Artifact Hub stop running vulnerability scans against projects?

**Comment by huntingcodes-001** [comment]
@leecalcote assign this task to me, i will surely be able to solve it!

I'll review the security report, prioritize vulnerabilities, and implement necessary fixes to improve the score

**Comment by Alianait** [comment]
@leecalcote Actually if report its not ready I can take it from here

**Comment by sangramrath** [comment]
@mack-droid there have been updates made to images and libraries since this issue was opened through other issues (and their PRs). I will check if this needs work and update.

**Comment by hortison** [comment]
These have long-since been upgraded. None of Meshery's container images carry these versions of dependencies any longer.
